### PR TITLE
chore(experimental): Add Ast Functor

### LINKED ISSUE #4954: Create Functor Ast to control interleaving
### Problem

Implementing https://github.com/noir-lang/noir/issues/4953 is a difficult task since it requires fundamentally changing the compiler's architecture.

Most notably, passes can accidentally break the lock-step constraint by recursing on an expression themselves as in:
```
fn type_check(expr: Expr) {
    match expr {
        Expr::Add(a, b) => {
            let a_type = type_check(a); // breaks lock-step! 
            let b_type = type_check(b); // we're now several nodes out of sync with other passes!
        }
        ...
    }
}
```
In this case, breaking lock step in the type checker would mean it is type checking nodes which haven't been name resolved yet, and would likely error as a result.

### Happy Case

We can add a Functor version of the Ast with all recursive Expression/Statement nodes replaced with a generic `T` argument. This is similar to the technique used in "recursion schemes" for reusable recursive helper functions, although Rust does not have the expressitivity to represent the normal Ast as a `Fix`ed version of the Functor version (if this means nothing to you, feel free to ignore this remark). For us, this just means it needs to be a separate enum from the normal Ast.

### Project Impact

None

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves #4954

## Summary\*

Adds a functor (mappable) version of our Ast enum.

## Additional Context

This is needed for interleaving three compiler passes for the metaprogramming work. See the linked issue and the issue linked from it as well.

The actual Ast isn't used at all yet but will eventually be used by the future composer struct.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by TomAFrench — CHANGES_REQUESTED** [review]
Blocking for now as I want to get a sync in before this is merged.

**Comment by jfecher** [comment]
> LGTM but hard to tell if anything's missed without tests

There's quite a bit still that will need to be changed/adapted from this Ast representation. E.g. if we want to hand this to type checking and interpretation those passes need DefinitionIds for names instead of just the actual Path. So really this structure should be a merger of the Ast and Hir somewhat. I'm leaving these changes for the PR which will actually use this ast in the composer though (we need a better name for this other than just the functor ast).

**Comment by jfecher** [comment]
Going to hold off on this PR while I re-evaluate if this is the direction I want to go with still.

**Pull request closed without merging**
